### ISSUE TITLE
./pyconfig.h:1438:0: warning: "_GNU_SOURCE" redefined [enabled by default]

### ISSUE THREAD

**Issue by 6ccb8c66-b6ca-4c37-b449-e35b11a6b9e4** [python/cpython#74980]
BPO | [30797](https://bugs.python.org/issue30797)
--- | :---
Nosy | @vstinner, @ned-deily, @segevfiner
PRs | <li>python/cpython#2615</li><li>python/cpython#2670</li><li>python/cpython#2671</li><li>python/cpython#2672</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/vstinner'
closed_at = <Date 2017-07-16.08:55:23.402>
created_at = <Date 2017-06-28.22:24:05.887>
labels = ['build', '3.7']
title = './pyconfig.h:1438:0: warning: "_GNU_SOURCE" redefined [enabled by default]'
updated_at = <Date 2017-07-16.08:55:23.401>
user = 'https://github.com/segevfiner'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-16.08:55:23.401>
actor = 'ned.deily'
assignee = 'vstinner'
closed = True
closed_date = <Date 2017-07-16.08:55:23.402>
closer = 'ned.deily'
components = ['Build']
creation = <Date 2017-06-28.22:24:05.887>
creator = 'Segev Finer'
dependencies = []
files = []
hgrepos = []
issue_num = 30797
keywords = []
message_count = 7.0
messages = ['297238', '297860', '297861', '297871', '297931', '298185', '298236']
nosy_count = 3.0
nosy_names = ['vstinner', 'ned.deily', 'Segev Finer']
pr_nums = ['2615', '2670', '2671', '2672']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = None
url = 'https://bugs.python.org/issue30797'
versions = ['Python 2.7', 'Python 3.3', 'Python 3.4', 'Python 3.5', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by 6ccb8c66-b6ca-4c37-b449-e35b11a6b9e4** [comment]
This is caused by \_GNU_SOURCE being defined here: https://github.com/python/cpython/blob/c08177a1ccad2ed0d50898c2731b518c631aed14/configure.ac#L124, But also getting defined by AC_USE_SYSTEM_EXTENSIONS (https://github.com/python/cpython/blob/c08177a1ccad2ed0d50898c2731b518c631aed14/configure.ac#L906).

AC_USE_SYSTEM_EXTENSIONS does use an #ifndef but Python's AC_DEFINE seems to be getting order after it.

**Comment by ned-deily** [comment]
New changeset 05b72ede95521b2d897cb4c7b034139b5437c592 by Ned Deily in branch '3.6':
bpo-30797, bpo-30694: Avoid \_GNU_SOURCE redefined warning in xmlparse.c (bpo-2615)
https://github.com/python/cpython/commit/05b72ede95521b2d897cb4c7b034139b5437c592

**Comment by ned-deily** [comment]
Thanks for the report.  This warning is currently visible on many/all of the buildbots on all branches.  The conflict is caused by the upgrade of the bundled to libexpat to version 2.2.1 (bpo-30694). 2.2.1 introduced a define of \_GNU_SOURCE into [Modules/expat/xmlparse.c](https://github.com/python/cpython/blob/main/Modules/expat/xmlparse.c).  The definition in xmlparse is simply:

```c
#define _GNU_SOURCE
```


But Python, in pyconfig.h, which ./configure generates from pyconfig.h.in, may already have defined \_GNU_SOURCE thusly:

```c
#define _GNU_SOURCE 1
```


pyconfig.h gets included into our copy of [Modules/expat/expat_config.h](https://github.com/python/cpython/blob/main/Modules/expat/expat_config.h) which is included later in xmlparse.c causing the compile warning.

The simplest way to avoid the warning is to just change the definition in xmlparse.c to match that in pyconfig.h; the compilers don't complain if the definition is identical.  I have now merged that fix into the 3.6 branch so builders of 3.6.2 won't see the warning.  I will leave this open for @haypo to review and to handle merging into the other branches affected by bpo-30694 (master, 3.5, 2.7, 3.4, 3.3).

**Comment by ned-deily** [comment]
One more thing: building with a recent gcc on Linux:
$ gcc --version
gcc (Debian 6.3.0-18) 6.3.0 20170516

also gives an "unused but set variable" warning.

```c
./Modules/expat/xmlparse.c: In function ‘gather_time_entropy’:
./Modules/expat/xmlparse.c:780:7: warning: variable ‘gettimeofday_res’ set but not used [-Wunused-but-set-variable]
   int gettimeofday_res;
       ^~~~~~~~~~~~~~~~
```


Since I didn't see it in time, I didn't try to fix this for 3.6.2 but it would be best to silence it going forward.

**Comment by ned-deily** [comment]
New changeset bdabd7666032ce356d550da21c35e4bee5b3448c by Ned Deily in branch '3.6':
bpo-30797, bpo-30694: Avoid \_GNU_SOURCE redefined warning in xmlparse.c (bpo-2615)
https://github.com/python/cpython/commit/bdabd7666032ce356d550da21c35e4bee5b3448c

**Comment by vstinner** [comment]
New changeset 884c4ca33ab84b9fc57338cf59f79e0436d3da43 by Victor Stinner (Segev Finer) in branch '2.7':
[2.7] bpo-30797: Avoid \_GNU_SOURCE redefined warning in xmlparse.c (GH-2670) (bpo-2672)
https://github.com/python/cpython/commit/884c4ca33ab84b9fc57338cf59f79e0436d3da43

**Comment by 6ccb8c66-b6ca-4c37-b449-e35b11a6b9e4** [comment]
The fix is now in 3.5-3.7 and 2.7, unless you intend to also back port to 3.4 and 3.3, this can be closed.
